### PR TITLE
Use timestamp instead of block number in predictoor

### LINKED ISSUE #779: Use timestamp instead of block number in predictoor
Use timestamp instead of block number in predictoor

Parent issue: [pdr-private:26](https://github.com/oceanprotocol/pdr-private/issues/26)

### PR DESCRIPTION
**Pull request by trizin**
Fixes #779

Changes proposed in this PR:

- Use timestamp instead of block number 
- Calculate everything based on epoch number, and the epoch number based on timestamp

### CODE REVIEW

File: contracts/templates/ERC20Template3.sol [hunk 1]
```diff
@@ -111,13 +111,14 @@ contract ERC20Template3 is
     mapping(uint256 => uint256) private roundSumStakes;
     mapping(uint256 => bool) public trueValues; // true values submited by owner
     mapping(uint256 => Status) public epochStatus; // status of each epoch
-    mapping(uint256 => uint256) private subscriptionRevenueAtBlock; //income registred
+    mapping(uint256 => uint256) private _subscriptionRevenueAtSlot; //income registred
```
Comment (alexcos20) [inline contracts/templates/ERC20Template3.sol:114]:
```suggestion
    mapping(uint256 => uint256) private subscriptionRevenueAtEpoch; //income registred
```

File: contracts/templates/ERC20Template3.sol [hunk 2]
```diff
@@ -961,24 +953,22 @@ contract ERC20Template3 is
...
     }
 
-    function getSubscriptionRevenueAtBlock(
-        uint256 blocknum
+    function getsubscriptionRevenueAtEpoch(
+        uint256 _epoch
     ) public view returns (uint256) {
-        uint256 slot = railBlocknumToSlot(blocknum);
-        return (subscriptionRevenueAtBlock[slot]);
+        return (_subscriptionRevenueAtSlot[_epoch]);
```
Comment (alexcos20) [inline contracts/templates/ERC20Template3.sol:967]:
```suggestion
        return (subscriptionRevenueAtEpoch[_epoch]);
```

File: contracts/templates/ERC20Template3.sol [hunk 3]
```diff
@@ -1032,64 +1020,62 @@ contract ERC20Template3 is
...
-        uint256 slot = railBlocknumToSlot(blocknum);
-        require(roundSumStakes[slot] == 0);
+    function redeemUnusedSlotRevenue(uint256 _epoch) external onlyERC20Deployer {
+        require(curEpoch() >= _epoch);
+        require(roundSumStakes[_epoch] == 0);
         require(feeCollector != address(0), "Cannot send fees to address 0");
         IERC20(stakeToken).safeTransfer(
             feeCollector,
-            subscriptionRevenueAtBlock[slot]
+            _subscriptionRevenueAtSlot[_epoch]
```
Comment (alexcos20) [inline contracts/templates/ERC20Template3.sol:1078]:
```suggestion
            subscriptionRevenueAtEpoch[_epoch]
```

File: contracts/templates/ERC20Template3.sol [hunk 4]
```diff
@@ -1166,38 +1149,32 @@ contract ERC20Template3 is
...
             if(num_epochs<1)
                 num_epochs=1;
             uint256 amt_per_epoch = amount / num_epochs;
-            // for loop and add revenue for blocksPerEpoch blocks
+            // for loop and add revenue for secondsPerEpoch blocks
             for (uint256 i = 0; i < num_epochs; i++) {
-                // TODO FIND A WAY TO ACHIEVE THIS WITHOUT A LOOP
-                subscriptionRevenueAtBlock[
-                    slot + blocksPerEpoch * (i)
+                _subscriptionRevenueAtSlot[
```
Comment (alexcos20) [inline contracts/templates/ERC20Template3.sol:1173]:
```suggestion
                subscriptionRevenueAtEpoch[
```

**Comment by alexcos20** [comment]
The concept of epoch is wrong.
Epoch  is timestamp of beginning of that epoch. 

epoch = (  ts / seconds_per_epoch) * seconds_per_epoch


Some math for 5m epoch:
 - ts:  1690702341 ->  epoch = 1690702200
 - ts:  1690702458  ->  epoch = 1690702200  (same epoch as above)
 - ts:  1690728701  ->  epoch = 1690728600

For 15m epochs:
 - ts:  1690702341 ->  epoch = 1690702200   (same epoch as above)
 - ts:  1690702458  ->  epoch = 1690702200  (same epoch as above)
 - ts:  1690728701  ->  epoch = 1690728300  (this has changed, since is a 15m epoch)
 
 
 This way, the flow is much cleaner:
  - I will predict for epoch 1690702200     (which in our case, is also the candle timestamp)
  - I will claim rewards for epoch 1690702200
  - etc..

Also, we can ditch start_timestamp, because we don't care if somebody will call getAggVal (epoch = 2), it will get None.  Or if somebody tries to claim rewards for epoch 100.

**Comment by alexcos20** [comment]
If we agree that would be beneficial to allow users to use timestamp as argument for all functions (claim, predict, submitTrueVal, etc), then the debate is basely what to use in the mappings, right ?

In the current implementation, we used railBlocknumToSlot.    We are going to use it again (railTsToEpoch), but instead of slots, let's use term epoch  -> we have the mappings:
 - roundSumStakesUp[epoch]
 - roundSumStakes[epoch]
 - trueValues[epoch]
 - etc..

So, question is:   Is epoch one of the following:
 - floor(timestamp / seconds per epoch)   (your proposal)
 - floor(timestamp / seconds_per_epoch) * seconds_per_epoch  (my proposal)
 
 
 I lean for the 2nd, for the following reasons:
 - no extra cognitive load (in your proposal, we are adding an additional term.  In main, it's candle timestamp, which is already something familiar for both predictors and traders
 - no extra code on subgraph
 - no extra code on frontend  (much cleaner to say:   prediction at 2022-01-01 13:00:00  then prediction for epoch 349787988

**Comment by trentmc** [comment]
> the debate is what to use for the mappings, right ?
+1

How about: Alex's proposal but we don't use "epoch" we use "epstart" (ie epoch start time). It's short yet clear.

**Comment by trentmc** [comment]
>  the need to validate the timestamp within every function that accepts it, leading to an overall increase in complexity. 

There's added complexity in having to validate it

And reduced complexity because we don't have to track some special new number

Overall, imo complexity is lower because we are reducing state and making the variables that we work with to be more readily human-understandable.

**Comment by alexcos20** [comment]
> I agree with the advantages of using timestamp. However, one potential drawback is the need to validate the timestamp within every function that accepts it, leading to an overall increase in complexity. Does it make sense to use epoch number for internal mappings etc. and automatically convert timestamp to epoch number in external functions or is it better to ditch epochs completely? By epoch, I refer to the number that increases by 1 per `seconds per epoch` seconds.

we are doing that now with railBlocknumToSlot  :)

**Comment by trentmc** [comment]
We (Berkay, Alex, Trent) agreed on
- Alex's proposal
- name "epoch_start"

**Review by alexcos20 — APPROVED** [review]
lgtm

**Review by trentmc — APPROVED** [review]

**Pull request merged**